osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorPlayer.cs

```diff
@@ -53,6 +53,15 @@ protected override void UpdateAfterChildren()
         }
 
         protected override GameplayClockContainer CreateGameplayClockContainer(WorkingBeatmap beatmap, double gameplayStart)
-            => new GameplayClockContainer(spectatorPlayerClock);
+        {
+            var gameplayClockContainer = new GameplayClockContainer(spectatorPlayerClock);
+
+            // Directionality is important, as BindAdjustments is... not actually a bidirectional bind...
+            // We want to ensure that any adjustments applied by the Player instance are applied to the SpectatorPlayerClock
+            // so they can be consumed by the spectator screen (and applied to the master clock / track).
+            spectatorPlayerClock.GameplayAdjustments.BindAdjustments(gameplayClockContainer.GameplayAdjustments);
```

Comment:
This is breaking my brain. So let me get this straight.
- The spectator's gameplay clock container has adjustments. These are coming from who knows where (I can't tell).
- That gameplay clock's adjustments are bound to the catchup clock (`spectatorPlayerClock`).
  - This is kind of weird, but maybe it's alright.
- Then multi-spectator binds the active audio sources' catchup clock to the master clock?
  - I don't get this one. I don't think that the active audio source should do anything with the track.

I can't handle this code, my brain-stack can't fit it all in one frame. Can this code be refactored first to avoid this and structure code in the way that intentions are clear, prior to the rest of this PR?

Reply:
> * The spectator's gameplay clock container has adjustments. These are coming from who knows where (I can't tell).

They are applied in `Player` (as they have always been):

https://github.com/ppy/osu/blob/cf6bb3b03037b128f3dbeb4e124d2750d9c5783c/osu.Game/Screens/Play/Player.cs#L1003-L1004

> * That gameplay clock's adjustments are bound to the catchup clock one level higher (`spectatorPlayerClock`).

They are not bound, just exposed.

> * Then multi-spectator, yet another level higher, binds the active audio sources' catchup clock to the master clock?

This is applying the gameplay adjustments (ie. the current state of `ModRamp` or `DoubleTime`) to the `MasterGameplayClock`, which is managing the `Track` internally.

> I can't handle this code, my brain-stack can't fit it all in one frame. Can this code be refactored first to avoid this and structure code in the way that intentions are clear, prior to the rest of this PR? I can't do this sort of multi level binding between parents and different children of parents.

I'm open to proposals. This is the best way I could figure out how to get the current players' adjustments applied to the `MasterGameplayClock` in the `SpectatorScreen`.

---

Unrelated, but as I mentioned previously in this PR thread, this only happened to work by chance until now. *Every* player was individually operating on the track, resetting then applying, resetting then applying their changes for each spectator screen.